### PR TITLE
@dace.program performance optimizations

### PR DESCRIPTION
**Pull request by tbennun**
- [x] Cache compiled SDFG from called `@dace.program`s
- [x] Support for transforming DacePrograms directly (#691) and calling SDFGs with their original Python argument order
- [x] More robust and safer checks for "if the SDFG is already loaded, don't override/recompile"
- [x] Minor auto-optimization fix
- [x] Return values are allocated as `np.empty` rather than `np.zeros`

### CODE REVIEW

File: dace/frontend/python/parser.py [hunk 1]
```diff
@@ -282,50 +248,99 @@ def __call__(self, *args, **kwargs):
...
 
+    def parse(self, *compilation_args, strict=None, save=False) -> SDFG:
+        """ 
+        Try to parse a DaceProgram object and return the `dace.SDFG` object
+        that corresponds to it.
+        :param function: DaceProgram object (obtained from the ``@dace.program``
+                        decorator).
+        :param compilation_args: Various compilation arguments e.g. dtypes.
+        :param strict: Whether to apply strict transformations or not (None
+                    uses configuration-defined value).
```
Comment (definelicht) [inline dace/frontend/python/parser.py:272]:
```suggestion
        :param strict: Whether to apply strict transformations or not (None
                       uses configuration-defined value). 
```

Comment (definelicht) [reply dace/frontend/python/parser.py:272]: I wish yapf could handle docstrings... There are [tools that can do this](https://github.com/myint/docformatter) (albeit separate), we could consider if it's worth adopting this at some point.

File: dace/frontend/python/parser.py [hunk 2]
```diff
@@ -282,50 +248,99 @@ def __call__(self, *args, **kwargs):
...
+            sdfg.apply_strict_transformations()
+
+            # Split back edges with assignments and conditions to allow richer
+            # control flow detection in code generation
+            xfh.split_interstate_edges(sdfg)
+
+        # Save the SDFG. Skip this step if running from a cached SDFG, as
+        # it might overwrite the cached SDFG.
+        if not Config.get_bool('compiler', 'use_cache') and save:
+            sdfg.save(os.path.join('_dacegraphs', 'program.sdfg'))
```
Comment (definelicht) [inline dace/frontend/python/parser.py:306]: I know this is not new, but out of curiosity -- why do we always save the SDFG here? It's already saved in the `.dacecache`.

File: dace/transformation/auto/auto_optimize.py [hunk 3]
```diff
@@ -166,17 +166,28 @@ def tile_wcrs(graph_or_subgraph: GraphViewType,
...
+
+                # Make transient only if there is one WCR/stream
+                if to_apply is not None:
+                    to_apply = None
+                    break
+
+                to_apply = (dataflow.StreamTransient, {},
+                            dict(tasklet=tasklet,
+                                 map_exit=mapexit,
+                                 outer_map_exit=outer_mapexit))
```
Comment (definelicht) [inline dace/transformation/auto/auto_optimize.py:190]: This snuck in? ;-)

**Review by definelicht — APPROVED** [review]
Approving because comments are so minor, just murge when ready

File: dace/frontend/python/parser.py (hunk 1 above)
Comment (tbennun) [reply dace/frontend/python/parser.py:272]: yeah no idea why yapf doesn't do this.

File: dace/frontend/python/parser.py (hunk 2 above)
Comment (tbennun) [reply dace/frontend/python/parser.py:306]: we now don't :) the diff that for some reason doesn't show up is that the argument is now `save=False` 💯

File: dace/transformation/auto/auto_optimize.py (hunk 3 above)
Comment (tbennun) [reply dace/transformation/auto/auto_optimize.py:190]: 1. it's part of the PR description
2. it appeared on auto-optimized `@dace.program`s

File: dace/frontend/python/parser.py (hunk 2 above)
Comment (definelicht) [reply dace/frontend/python/parser.py:306]: Ah, gr8

**Pull request merged**